### ISSUE TITLE
`front` and `tail` for `AbstractVector`

### ISSUE THREAD

**Issue by jariji** [JuliaLang/julia#53045]
```jl
Base.front(x::AbstractVector) = x[begin:end-1]
Base.tail(x::AbstractVector) = x[begin+1:end]
```

Currently these are only defined for `Tuple` but it would be nice to have them for `AbstractVector` too.

**Comment by nandoconde** [comment]
I think this cannot be defined in general with  `-1` and `+1` because of `OffsetArrays`, right?

**Comment by barucden** [comment]
We have `prevind` and `nextind` defined for `AbstractArray`s, so a general way might be to use  `nextind(firstindex(x))` to get the second index, and `prevind(lastindex(x))` to get the second to last index. Nevertheless, the `prevind` and `nextind` methods are defined with a fixed step size of 1:
https://github.com/JuliaLang/julia/blob/0eec0317b1991c477b87d1760233bed1043e052d/base/abstractarray.jl#L212-L213

(also #42084)

**Comment by MasonProtter** [comment]
Almost all abstract array code would break if you used a type with non-zero steps.

**Comment by MasonProtter** [comment]
> ```julia
> Base.front(x::AbstractVector) = x[begin:end-1]
> Base.tail(x::AbstractVector) = x[begin+1:end]
> ```

Should probably be a view.

**Comment by MasonProtter** [comment]
Oops, yeah I did.

**Comment by jishnub** [comment]
I feel this should work the other way. For tuples, an indexing operation like `x[2:end]` calls `Base.tail` internally, so instead of specializing `front` and `tail`, we should be encouraging indexing operations for both tuples and vectors.

**Comment by jariji** [comment]
@MasonProtter 
> Should probably be a view.

From casual inspection, I think most of the [array functions](https://docs.julialang.org/en/v1/base/arrays/) are not views. @aplavin's FlexiMaps.jl defines `mapview(f, xs)` which has "view" in its name while `Base.map` allocates a new array.

What justifies making `Base.tail` a view?

**Comment by jariji** [comment]
@aplavin 

> surely `tail` shouldn't be a view for consistency

If you mean consistency with `tail(::Tuple)::Tuple`, it's true that it returns a materialized value. However, a view of a tuple isn't a well defined concept in Julia afaik and I'm not sure how the would differ from the regular tuple. So perhaps whether `tail` returns a view is still an open question.

**Comment by mbauman** [comment]
More concretely, tuples are immutable so it doesn't matter — and isn't even observable — whether they're views or not.

A the functions that I'd expect to behave similarly are `first(A, n)` and `last(A, n)`; those create copies.  Perhaps this should just be proxies for those.

```julia
Base.front(x::AbstractVector) = first(x, length(x)-1)
Base.tail(x::AbstractVector) = last(x, length(x)-1)
```

**Comment by adienes** [comment]
btw what do we want these methods for in the first place? I mean I'm not opposed, per se, but it's not entirely clear to me what kind of function wants to use these to be generic over tuple and array, and isn't able to use `getindex`

**Comment by jariji** [comment]
@adienes 

> btw what do we want these methods for in the first place? I mean I'm not opposed, per se, but it's not entirely clear to me what kind of function wants to use these to be generic over tuple and array, and isn't able to use getindex

In the case where `front` and `tail` are equivalent to `getindex` (e.g. `Vector`), there is a syntactic benefit: one-argument functions like can be used point-free, with `|>` or `∘`, which is sometimes nicer than having to give a name to the data argument and using `getindex`.


In other cases, the iteration order is not the same as the `getindex` indices, and the semantics of `front`/`tail` would use iteration order not `getindex` order. For example:


```jl
using Dictionaries
julia> d = Dictionary([2, 1], ['b', 'a'])
2-element Dictionary{Int64, Char}:
 2 │ 'b'
 1 │ 'a'

julia> d[1]
'a': ASCII/Unicode U+0061 (category Ll: Letter, lowercase)

julia> first(d, 1)
1-element Vector{Char}:
 'b': ASCII/Unicode U+0062 (category Ll: Letter, lowercase)
```

**Comment by adienes** [comment]
> front/tail would use iteration order not getindex order. For example

wait, what? this I maybe don't understand. when would it be desirable for `arr[begin+1:end]` to not be the same order of elements as `collect(Iterators.rest(arr, firstindex(arr)+1))`

**Comment by jariji** [comment]
I agree the invariant `arr[begin+1:end] == collect(Iterators.rest(arr, firstindex(arr)+1))` always holds for `arr::AbstractVector`.

This issue is focused on `AbstractVector` in particular, but I think it's useful to keep in mind a hypothetical generic contract for `front` whose semantics aren't specific to a particular type. Eg `Dictionaries.Dictionary` has specified iteration order (unlike `Dict`), and that iteration order is not consistent with its `getindex` order (unlike `AbstractVector`). I'm claiming `front`'s generic contract would be defined in terms of iteration order rather than its `getindex` order.

**Comment by mcabbott** [comment]
Nobody has mentioned `Iterators.peel` so far, which is a generic way of splitting first from the rest, and clearly uses iteration order. 

What types it returns could probably be changed if desired... for instance to to return `first(t), tail(t)` on a Tuple, and perhaps `v[1], @view v[2:end]` on a Vector.

It doesn't satisfy the desire to have a standalone function for `tail`. I'm surprised that `Iterators.rest` has an (undocumented?) 1-arg method `rest(itr) = itr`, otherwise that would be a logical name for `last∘Iterators.peel`, which returns an Iterators.Rest.

**Comment by jariji** [comment]
@mcabbott 
> I'm surprised that Iterators.rest has an (undocumented?) 1-arg method rest(itr) = itr, otherwise that would be a logical name for last∘Iterators.peel, which returns an Iterators.Rest.

Yeah. Unfortunately, unlike `Iterators.rest`, `Base.rest(collection[, state])` is documented.  https://github.com/JuliaLang/julia/issues/49274

**Comment by MagicMuscleMan** [comment]
And while we are at it: Any reason not to define these methods for `AbstractString`, too?

I think whenever sensibly possible, all these functions should have a method for `Tuple`, `AbstractString` and `AbstractVector` to reduce mental complexity.

We can create another issue, however, if you don't want to combine it in the same PR.
